StitchCore/StitchCoreTests/TestHarness/TestHarness.swift

```diff
@@ -0,0 +1,172 @@
+    }
+}
+
+final class TestHarness {
+    var adminClient: StitchAdminClient!
+    let initPromise: Promise<Void>
+    let serverUrl: String
+    let username: String
+    let password: String
+    var testApp: AppView?
+    var stitchClient: StitchClient?
+    var userCredentials: (username: String, password: String)?
+    var groupId: String?
+    var user: UserView?
+
+    lazy var apps: AppsEndpoint = self.adminClient.apps(withGroupId: self.groupId!)
+
+    var app: AppEndpoint {
+        guard let testApp = self.testApp else {
+            fatalError("App must be created first")
+        }
+        return self.apps.app(withAppId: testApp.id)
+    }
+
+    fileprivate init(username: String = "unique_user@domain.com",
+                     password: String = "password",
+                     serverUrl: String = defaultServerUrl) {
+        self.username = username
+        self.password = password
+        self.serverUrl = serverUrl
+        self.initPromise = Promise()
+        initPromise.then {
+            StitchAdminClientFactory.create(baseUrl: self.serverUrl)
+        }.done {
+            self.adminClient = $0
+        }.cauterize()
+    }
+
+    func teardown() -> Promise<Void> {
+        return self.app.remove().asVoid()
+    }
+
+    func authenticate() -> Promise<Void> {
+        return self.adminClient.authenticate(
+            provider: EmailPasswordAuthProvider.init(username: self.username,
+                                                     password: self.password)
+        ).then { (_) -> Promise<UserProfile> in
+            return self.adminClient.fetchUserProfile()
+        }.done { (userProfile: UserProfile) in
+            self.groupId = userProfile.roles!.first?.groupId
+        }
+    }
+
+    func createApp(testAppName: String = "test-\(ObjectId.init().hexString)") -> Promise<AppView> {
+        return self.apps.create(name: testAppName).flatMap {
+            self.testApp = $0
+            return $0
+        }
+    }
+
+    func createUser(email: String = "test_user@domain.com",
+                    password: String = "password") -> Promise<UserView> {
+        self.userCredentials = (username: email, password: password)
+        return self.app.users.create(data: UserCreator.init(email: email, password: password)).flatMap {
+            self.user = $0
+            return self.user
+        }
+    }
+
+    func add(serviceConfig: ServiceConfigs, withRules rules: Rule...) -> Promise<ServiceView> {
+        return self.app.services.create(data: serviceConfig).then { view in
+            return when(resolved: rules.map {
+                return self.app.services.service(withId: view.id).rules.create(data: $0)
+            }).flatMap { _ in view }
+        }
+    }
+    
+    func add(providerConfig: ProviderConfigs) -> Promise<AuthProviderView> {
+        return self.app.authProviders.create(data: providerConfig)
+    }
+
+    func addDefaultUserpassConfig() -> Promise<AuthProviderView> {
```

Comment:
[nit] We're not adding a config but adding a provider (we just provide a config).

Reply:
👍